### PR TITLE
Add more optimisation rules for shifts

### PR DESCRIPTION
**Pull request by axic**
Part of #6015.

### CODE REVIEW

File: libevmasm/RuleList.h [hunk 1]
```diff
@@ -140,6 +140,10 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleListPart2(
 		{{Instruction::MOD, {0, X}}, [=]{ return u256(0); }, true},
 		{{Instruction::EQ, {X, 0}}, [=]() -> Pattern { return {Instruction::ISZERO, {X}}; }, false },
 		{{Instruction::EQ, {0, X}}, [=]() -> Pattern { return {Instruction::ISZERO, {X}}; }, false },
+		{{Instruction::SHL, {X, 0}}, [=]{ return X; }, false},
+		{{Instruction::SHR, {X, 0}}, [=]{ return X; }, false},
+		{{Instruction::SHL, {0, X}}, [=]{ return u256(0); }, false},
```
Comment (chriseth) [inline libevmasm/RuleList.h]:
```suggestion
		{{Instruction::SHL, {0, X}}, [=]{ return u256(0); }, true},
```

File: libevmasm/RuleList.h [hunk 2]
```diff
@@ -140,6 +140,10 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleListPart2(
...
+		{{Instruction::SHR, {0, X}}, [=]{ return u256(0); }, false},
```
Comment (chriseth) [inline libevmasm/RuleList.h]:
```suggestion
		{{Instruction::SHR, {0, X}}, [=]{ return u256(0); }, true},
```

**Review by chriseth — CHANGES_REQUESTED** [review]
The bool at the end means "potentially removes parts of the expression tree that might have side-effects". It might be that the documentation about that got lost during the refactorings.

**Comment by erak** [comment]
The documentation in `SimplificationRule.h` could updated, I guess. The behavior @chriseth described is not that obvious when looking at the code / docs.

**Comment by axic** [comment]
Fixed.

File: libevmasm/RuleList.h [hunk 3]
```diff
@@ -140,6 +140,10 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleListPart2(
 		{{Instruction::MOD, {0, X}}, [=]{ return u256(0); }, true},
 		{{Instruction::EQ, {X, 0}}, [=]() -> Pattern { return {Instruction::ISZERO, {X}}; }, false },
 		{{Instruction::EQ, {0, X}}, [=]() -> Pattern { return {Instruction::ISZERO, {X}}; }, false },
+		{{Instruction::SHL, {X, 0}}, [=]{ return X; }, false},
```
Comment (chriseth) [inline libevmasm/RuleList.h]: this shifts a zero by x bits and thus should be replaced by zero.

**Comment by axic** [comment]
Seems the operands are reverse of shl/shr compared to sub/div. Fixed.

**Review by chriseth — APPROVED** [review]

**Pull request merged**